File: Maanfee.JsInterop/wwwroot/js/Dom.js
/*
 * JavaScript Library 0.2.0
 *
 *
 *
 * Date: 01-11-01
 */

// ========== DOM Manipulation ==========

export function querySelector(selector) {
    return document.querySelector(selector);
}

export function querySelectorAll(selector) {
    return Array.from(document.querySelectorAll(selector));
}

function _Selector(selector, selectors) {
    let elements = [];

    if (selector && selector !== "") {
        // Use querySelector - single element
        const element = querySelector(selector);
        if (element) {
            elements.push(element);
        }
    }
    else if (selectors && selectors !== "") {
        // Use querySelectorAll - multiple elements
        elements = querySelectorAll(selectors);
    }

    return elements;
}

// ========== Content Manipulation ==========

export function Text(selector, selectors, Content = null) {

    let elements = _Selector(selector, selectors);
    if (elements.length === 0) {
        return null;
    }

    if (Content === null) {
        // GET
        elements = Array.from(elements);
        return JSON.stringify(elements.map(item => item.textContent || item.innerText || ''));
    } else {
        // SET 
        elements.forEach(item => {
            item.innerHTML = Content;
        });
        return Content;
    }
}

export function HTML(selector, selectors, Content = null) {

    let elements = _Selector(selector, selectors);
    if (elements.length === 0) {
        return null;
    }

    if (Content === null) {
        // GET
        elements = Array.from(elements);
        return JSON.stringify(elements.map(item => item.innerHTML));
    } else {
        // SET 
        elements.forEach(item => {
            item.innerHTML = Content;
        });
        return Content;
    }
}

export function Val(selector, selectors, val = null) {

    let elements = _Selector(selector, selectors);
    if (elements.length === 0) {
        return null;
    }

    if (val === null) {
        // GET 
        elements = Array.from(elements);
        return JSON.stringify(elements.map(item => item.value));
    } else {
        // SET 
        elements.forEach(item => {
            item.value = val;
        });
        return val;
    }
}

export function Attr(selector, selectors, attributeName = null, attributeValue = null) {

    let elements = _Selector(selector, selectors);
    if (elements.length === 0) {
        return null;
    }

    if (attributeValue === null) {
        // GET
        elements = Array.from(elements);
        return JSON.stringify(elements.map(item => item.getAttribute(attributeName)));
    } else {
        // SET 
        elements.forEach(item => {
            item.setAttribute(attributeName, attributeValue);
        });
        return attributeValue;
    }
}

export function AddClass(selector, selectors, className) {
    let elements = _Selector(selector, selectors);
    if (elements.length === 0) {
        return null;
    }

    elements.forEach(el => el.classList.add(className));
    return className;
}

export function RemoveClass(selector, selectors, className) {
    let elements = _Selector(selector, selectors);
    if (elements.length === 0) return null;

    elements.forEach(el => el.classList.remove(className));
    return className;
}

export function ToggleClass(selector, selectors, className) {
    let elements = _Selector(selector, selectors);
    if (elements.length === 0) {
        return null;
    }

    elements.forEach(el => el.classList.toggle(className));
    return className;
}

export function Css(selector, selectors, propertyName, val = null) {
    let elements = _Selector(selector, selectors);
    if (elements.length === 0) {
        return null;
    }

    if (val === null) {
        // GET
        const values = elements.map(el => {
            const inline = el.style[propertyName];
            return inline || getComputedStyle(el)[propertyName] || "";
        });

        // همیشه آرایه JSON برگردون (حتی اگه یه المان باشه)
        return JSON.stringify(values);
    } else {
        // SET
        elements.forEach(item => {
            item.style[propertyName] = val;
        });
        return val;
    }
}

// ========== Events ==========

export function OnClick(selector, selectors) {

    let elements = _Selector(selector, selectors);
    if (elements.length === 0) {
        return null;
    }

    elements.forEach(item => {
        item.click();
    });
}

// ========== Event Listeners ==========

const _eventListeners = new Map();

export function AddEventListener(selector, selectors, eventName, dotNetHelper) {
    const elements = _Selector(selector, selectors);
    if (elements.length === 0) return;

    const key = `${selector || selectors}_${eventName}`;

    // حذف قبلی اگه وجود داشت
    if (_eventListeners.has(key)) {
        RemoveEventListener(selector, selectors, eventName);
    }

    const handler = (e) => {
        const target = e.target;
        const targetSelector = selector || selectors;

        dotNetHelper.invokeMethodAsync('OnEventFired', eventName, targetSelector);
    };

    elements.forEach(el => {
        el.addEventListener(eventName, handler);
    });

    _eventListeners.set(key, { elements, eventName, handler });
}

export function RemoveEventListener(selector, selectors, eventName) {
    const key = `${selector || selectors}_${eventName}`;
    const stored = _eventListeners.get(key);
    if (!stored) return;

    const { elements, handler } = stored;
    elements.forEach(el => {
        el.removeEventListener(eventName, handler);
    });

    _eventListeners.delete(key);
}

File: Maanfee.JsInterop/API/Dom.cs
using Microsoft.JSInterop;
using System.Text.Json;

namespace Maanfee.JsInterop
{
    public partial class Dom : JsService, IAsyncDisposable
    {
        public Dom(IJSRuntime JSRuntime) : base(JSRuntime, "Maanfee.JsInterop", "Dom.js")
        {
            //_ = Task.Run(async () => await EnsureModuleLoaded());
        }

        public new async ValueTask DisposeAsync()
        {
            if (!IsDisposed)
            {
                if (_Module != null)
                {
                    await _Module.InvokeVoidAsync("dispose");
                    await _Module.DisposeAsync();
                }

                await base.DisposeAsync();
            }
        }

        // ********************************************

        private string selector { get; set; }

        private string selectors { get; set; }

        // ========== DOM Selection ==========

        #region - DOM Selector -

        public async Task<Dom> QuerySelector(string Selector)
        {
            selector = Selector;
            selectors = null;
            return this;
        }

        public async Task<Dom> QuerySelectorAll(string Selectors)
        {
            selectors = Selectors;
            selector = null;
            return this;
        }

        #endregion

        // ========== Content Manipulation ==========

        #region - Test -

        public async Task<T> TextAsync<T>()
        {
            var data = await _Module.InvokeAsync<string>("Text", selector, selectors, null);

            if (!string.IsNullOrEmpty(data))
            {
                var list = JsonSerializer.Deserialize<List<string>>(data);

                if (typeof(T) == typeof(string))
                    return (T)(object)(list.FirstOrDefault() ?? string.Empty);
                else if (typeof(T) == typeof(IList<string>))
                    return (T)(object)list;
            }

            return default(T);
        }

        public async Task<Dom> TextAsync(string Text)
        {
            await _Module.InvokeAsync<string>("Text", selector, selectors, Text);
            return this;
        }

        #endregion

        #region - HTML -

        public async Task<T> HTMLAsync<T>()
        {
            var data = await _Module.InvokeAsync<string>("HTML", selector, selectors, null);

            if (!string.IsNullOrEmpty(data))
            {
                var list = JsonSerializer.Deserialize<List<string>>(data);

                if (typeof(T) == typeof(string))
                    return (T)(object)(list.FirstOrDefault() ?? string.Empty);
                else if (typeof(T) == typeof(IList<string>))
                    return (T)(object)list;
            }

            return default(T);
        }

        public async Task<Dom> HTMLAsync(string Text)
        {
            await _Module.InvokeAsync<string>("HTML", selector, selectors, Text);
            return this;
        }

        #endregion

        #region - Val -

        public async Task<T> ValAsync<T>()
        {
            var data = await _Module.InvokeAsync<string>("Val", selector, selectors, null);

            if (!string.IsNullOrEmpty(data))
            {
                var list = JsonSerializer.Deserialize<List<string>>(data);

                if (typeof(T) == typeof(string))
                    return (T)(object)(list.FirstOrDefault() ?? string.Empty);
                else if (typeof(T) == typeof(IList<string>))
                    return (T)(object)list;
            }

            return default(T);
        }

        public async Task<Dom> ValAsync(string Val)
        {
            await _Module.InvokeAsync<string>("Val", selector, selectors, Val);
            return this;
        }

        #endregion

        #region - Attr -

        // except hasAttribute(name) | removeAttribute(name) 

        public async Task<T> AttrAsync<T>(string AttributeName)
        {
            var data = await _Module.InvokeAsync<string>("Attr", selector, selectors, AttributeName, null);

            if (!string.IsNullOrEmpty(data))
            {
                var list = JsonSerializer.Deserialize<List<string>>(data);

                if (typeof(T) == typeof(string))
                    return (T)(object)(list.FirstOrDefault() ?? string.Empty);
                else if (typeof(T) == typeof(IList<string>))
                    return (T)(object)list;
            }

            return default(T);
        }

        public async Task<Dom> AttrAsync(string AttributeName, string AttributeValue)
        {
            await _Module.InvokeAsync<string>("Attr", selector, selectors, AttributeName, AttributeValue);
            return this;
        }

        #endregion

        #region - ClassList -

        public async Task<Dom> AddClassAsync(string className)
        {
            await _Module.InvokeAsync<string>("AddClass", selector, selectors, className);
            return this;
        }

        public async Task<Dom> RemoveClassAsync(string className)
        {
            await _Module.InvokeAsync<string>("RemoveClass", selector, selectors, className);
            return this;
        }

        public async Task<Dom> ToggleClassAsync(string className)
        {
            await _Module.InvokeAsync<string>("ToggleClass", selector, selectors, className);
            return this;
        }

        public async Task<bool> HasClassAsync(string className)
        {
            var result = await _Module.InvokeAsync<string>("HasClass", selector, selectors, className);
            return result == "true";
        }

        public async Task<IList<bool>> HasClassAllAsync(string className)
        {
            var data = await _Module.InvokeAsync<string>("HasClass", selector, selectors, className);
            if (string.IsNullOrEmpty(data))
                return new List<bool>();

            var list = JsonSerializer.Deserialize<List<string>>(data);
            return list.Select(x => x == "true").ToList();
        }

        #endregion

        #region - CSS -

        public async Task<Dom> CssAsync(string PropertyName, string Value)
        {
            await _Module.InvokeAsync<string>("Css", selector, selectors, PropertyName, Value);
            return this;
        }

        public async Task<T> CssAsync<T>(string PropertyName)
        {
            var data = await _Module.InvokeAsync<string>("Css", selector, selectors, PropertyName);

            if (!string.IsNullOrEmpty(data))
            {
                var list = JsonSerializer.Deserialize<List<string>>(data);

                if (typeof(T) == typeof(string))
                    return (T)(object)(list.FirstOrDefault() ?? string.Empty);
                else if (typeof(T).IsGenericType && typeof(T).GetGenericTypeDefinition() == typeof(List<>)
                      || typeof(T) == typeof(IList<string>) || typeof(T) == typeof(IEnumerable<string>))
                    return (T)(object)list;
            }

            return default(T);
        }

        #endregion

        // برای راحتی: متدهای اختصاصی برای display, opacity و ...
        //public async Task<Dom> ShowAsync() => await CssAsync("display", "block");
        //public async Task<Dom> HideAsync() => await CssAsync("display", "none");
        //public async Task<Dom> FadeInAsync(int durationMs = 300) => await CssAsync("transition", $"opacity {durationMs}ms").CssAsync("opacity", "1");
        //public async Task<Dom> FadeOutAsync(int durationMs = 300) => await CssAsync("transition", $"opacity {durationMs}ms").CssAsync("opacity", "0");


        // ==========================================

        #region - Events -

        public async Task<Dom> ClickAsync()
        {
            await _Module.InvokeAsync<Task>("OnClick", selector, selectors);
            return this;
        }

        #endregion

        #region - Events (addEventListener with callback) -

        private readonly Dictionary<string, DotNetObjectReference<DomEventHelper>> _eventHelpers = new();

        public async Task<Dom> OnAsync(string eventName, Func<EventArgs, Task> callback)
        {
            var helper = new DomEventHelper(this);
            var dotNetRef = DotNetObjectReference.Create(helper);

            var key = $"{selector ?? selectors}_{eventName}";

            // حذف قبلی اگر وجود داشت
            if (_eventHelpers.TryGetValue(key, out var oldRef))
            {
                oldRef.Dispose();
                _eventHelpers.Remove(key);
            }

            _eventHelpers[key] = dotNetRef;
            helper.Callback = callback;

            await _Module.InvokeVoidAsync("AddEventListener", selector, selectors, eventName, dotNetRef);

            return this;
        }

        public async Task<Dom> OffAsync(string eventName)
        {
            var key = $"{selector ?? selectors}_{eventName}";
            if (_eventHelpers.TryGetValue(key, out var dotNetRef))
            {
                await _Module.InvokeVoidAsync("RemoveEventListener", selector, selectors, eventName);
                dotNetRef.Dispose();
                _eventHelpers.Remove(key);
            }
            return this;
        }

        #endregion

        #region -  -



        #endregion

    }
}
